File: Problem2/go/problem2.go
package main

import (
  matrix "github.com/skelterjohn/go.matrix"
  "fmt"
  "math"
)

var relation = []float64{4,1,1,0}
var relation_mat = matrix.MakeDenseMatrix(relation, 2,2)
var initial_seq_values = []float64{8,2}
var init_mat = matrix.MakeDenseMatrix(initial_seq_values, 2,1)
var dict = make(map[float64]*matrix.DenseMatrix,0)
var pow = []float64{1,  -1}

/*************************************************************************
function calculate_even_fib_until:

Input: limit Integer

Output:
Computes the Even_Fibanacci_Series elements F(n) and F(n+1) where:
    F(n) < limit
    F(n+1) >= limit

> Even_Fibanacci_Series:
  It is the series formed by even elements of Fibonacci Series:
                2, 8, 34, 144, 610, ...
  It has a recurrence relation which be derived easily:
  |4, 1|  * | F(n) | = |F(n+1)|
  |1, 0|    |F(n-1)|   | F(n) |
  
Method:
Taking benefit of this recurrence relation, we use the matrix {4,1;1,0}
to exponentiate the matrix and calculate the power for which F(n) is
smaller than limit and F(n+1) is greater than limit. The latter part 
is tricky. First we exponentiate the matrix: 1, 2, 4, 8, 16 ... until 
a matrix appears for which F(n) is greater than the limit. Then we 
BINARY search in between the two powers after we get F(n) < limit and 
F(2*n) > limit.

Note: There are two modes of this function, in first mode the power m
of the matrix is exponentiated uptil m < limit and 2*m > limit. In 
second mode, it binary searches the matrix which gives the correct 
elements F(n) and F(n+1).

Complexity:
The complexity of this method is O(log n). Which is lesser than the 
linear search method which has a complexity of O(n).
*************************************************************************/

func calculate_even_fib_until(limit int64) (m []float64) {
  if limit <= 2 {         // base case: This cannot be determined by the recurrence relation
    m = []float64{2,0}
    return
  } 
  done := false
  var a,b,c *matrix.DenseMatrix
  a = relation_mat              // a holds older b before multiplication
  b = relation_mat              // b holds result of matrix multiplication
  c = relation_mat              // c holds matrix that is to be multiplied in current iteration
  dict[float64(1)] = relation_mat
  
  for !done {
    a = b
    b,_ = b.TimesDense(c)
    t, _ := b.TimesDense(init_mat)    // calculate resultant vector for a certain power of a matrix
    if t.ColCopy(0)[1] < float64(limit) {
      if pow[1] == -1 {     // pow[1] = -1 means we are exponentiating the matrix (first mode).
        pow[0] = 2*pow[0]
        c = b
        dict[pow[0]] = b
      } else {
        pow[0] = math.Ceil((pow[1] + pow[0])/float64(2))
        c = dict[math.Ceil((pow[1] - pow[0])/float64(2))]
      }
    } else {
      if pow[1] == -1 {
        pow[1] = pow[0]
        pow[0] = pow[0]/2
        dict[pow[1]] = b
        if pow[1] - pow[0] < 2 {
          c = dict[float64(1)]
        } else {
          c = dict[math.Ceil((pow[1] - pow[0])/float64(2))]
        }
        b = a
      } else {
        pow[1] = math.Ceil((pow[1] + pow[0])/float64(2))
        c = dict[math.Ceil((pow[1] - pow[0])/float64(2))]
        b = a
      }
    }
    t, _ = b.TimesDense(init_mat)
    if pow[1] != -1 && t.ColCopy(0)[1] < float64(limit) && t.ColCopy(0)[0] >= float64(limit) {
      done = true
      m = t.ColCopy(0)
    }
  }
  return
}

func find_sum(x int64) int64 {
  pow = []float64{1,  -1}                         // pow[1] = -1 means that we are in the process of exponentiation,
                                                  // otherwise, binary search
  f := calculate_even_fib_until(x)
  return int64(f[0] + f[1] - float64(2))/int64(4)
}

func main(){
  fmt.Println(find_sum(4000000))
}

File: Problem2/haskell/problem2.hs

import qualified Data.Matrix as Matrix
import qualified Data.Map as Map
import Data.Maybe

relation_mat = Matrix.fromList 2 2 [4,1,1,0]
init_mat = Matrix.fromList 2 1 [8,2]

-------------------------------------------------------
-- compute_vector calculates the calculates a vector,
-- given a relation matrix 'x' by multiplying it with
-- 'init_mat' [8,2]. This function is to compute nth
-- element of the even fibonacci series.
-------------------------------------------------------
compute_vector :: Matrix.Matrix Integer -> [Integer]
compute_vector x = Matrix.toList(x * init_mat)


--------------------------------------------------------------------
-- exponentiate_until:
-- Input:
--    -> limit  Integer
--    -> pow    Integer
--    -> y      Matrix.Matrix Integer
--    -> z      Map Integer (Matrix Integer)
-- y is the last computed relation matrix
-- z is a dictionary in which all the computed powers of initial 
-- matrix are stored and accessed from. pow shows to which power
-- we have exponentiated. It is used as key in z. Value in a pair in
-- z is a matrix of Integer(s), which is computed by exponentiating,
-- with the previous matrix 'y' from the dictionary.
-- Output:
-- It recursively exponentiates the intial matrix until a value in 
-- the series is reached which: fn > limit
--------------------------------------------------------------------

exponentiate_until ::  Integer -> Integer -> Matrix.Matrix Integer -> Map.Map (Integer) (Matrix.Matrix Integer) -> Map.Map (Integer) (Matrix.Matrix Integer)
exponentiate_until limit pow y z
  | fn > limit = z3
  | otherwise = exponentiate_until limit p2 y2 z2
  where y2 = y*y
        fn = last(compute_vector(y2))
        p2 = 2*pow
        z2 = Map.insert pow y z
        z3 = Map.insert p2 y2 z2
--------------------------------------------------------------------



--------------------------------------------------------------------
-- binary_search gets the value for which
--    >> f(n) < limit && f(n+1) >= limit
-- by searching between two powers provided in array 'pow'.
-- It also computes new powers of matrices between exponentiated 
-- powers in the 'dict'.
--------------------------------------------------------------------

binary_search :: [Integer] -> Map.Map (Integer) (Matrix.Matrix Integer) -> Integer -> [Integer]
binary_search pow dict limit
  | (last f_n < limit && head f_n >= limit) = f_n
  | (last f_nhalf < limit) = binary_search [p+p_half, p2] new_dict limit
  | otherwise = binary_search [p, p+p_half] new_dict limit
  where p = head pow
        p2 = last pow
        m_n = snd $ Map.elemAt (fromJust $ Map.lookupIndex p dict) dict
        f_2n = compute_vector $ snd $ Map.elemAt (fromJust $ Map.lookupIndex p2 dict) dict
        f_n = compute_vector $ m_n
        p_half = toInteger $ ceiling $ fromIntegral (p2-p)/2
        m_nhalf = m_n * (snd $ Map.elemAt (fromJust $ Map.lookupIndex p_half dict) dict)
        f_nhalf = compute_vector $ m_nhalf
        new_dict = Map.insert (p + p_half) m_nhalf dict
--------------------------------------------------------------------



--------------------------------------------------------------------
-- find_sum_even_fibonacci calculates sum of even fibonacci values
-- which are less than 'limit'. It first exponentiates and then 
-- binary searches between f(m) and f(2m), such that f(n) < limit 
-- and f(n+1) >= limit and m < n <= 2m.
--------------------------------------------------------------------
find_sum_even_fibonacci :: Integer -> Integer
find_sum_even_fibonacci limit
  | limit <= 2 = 0
  | otherwise = s
  where new_dict = Map.empty :: Map.Map (Integer) (Matrix.Matrix Integer)
        dict = exponentiate_until limit 1 relation_mat new_dict
        max_power = fst $ Map.findMax dict
        max_power_by_2 = toInteger $ ceiling $ fromIntegral max_power/2
        fn_and_fn_plus_1 = binary_search [max_power_by_2, max_power] dict limit
        s = toInteger $ ceiling $ fromIntegral (sum(fn_and_fn_plus_1)- 2)/4
--------------------------------------------------------------------


  
main = do
  let limit = 4000000
  putStrLn $ show $ find_sum_even_fibonacci limit